### PR TITLE
Type traits2

### PR DESCRIPTION
**Pull request by atkurtul**
## Description
Continuation of #546

### CODE REVIEW

File: include/nbl/builtin/hlsl/mpl.hlsl [hunk 1]
```diff
@@ -12,8 +12,6 @@ namespace nbl
 {
 namespace hlsl
 {
-namespace mpl
-{
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/mpl.hlsl:16]: do not remove the `mpl` namespace!

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 2]
```diff
@@ -34,7 +34,21 @@
...
+        cls       is object
+        cls&      is not an object
+        cls*      is object
+        int()     is not an object
+        int(*)()  is object
+        int(&)()  is not an object
+
   template<class T> struct is_scalar; (DONE)
   template<class T> struct is_compound; (DONE)
   template<class T> struct is_member_pointer; (TODO)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:54]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346099612

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/type_traits.hlsl:54]: `is_final` needs a comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346099913

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 3]
```diff
@@ -59,8 +73,8 @@
 
   // type property queries
   template<class T> struct alignment_of; (TODO)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:75]: add the specializations of `alignment_of` to `NBL_REGISTER_OBJ_TYPE`

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 4]
```diff
@@ -457,29 +497,28 @@ struct scalar_type<matrix<T,N,M> >
...
-NBL_NAMESPACE_HLSL_TYPE_TRAITS_BEGIN namespace impl { \
+NBL_NAMESPACE_HLSL_BEGIN namespace impl { \
 template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> { using type = T; }; \
-} NBL_NAMESPACE_HLSL_TYPE_TRAITS_END
+} NBL_NAMESPACE_HLSL_END
 
-#define typeid(expr) (sizeof(::nbl::hlsl::type_traits::impl::encode_typeid(expr))/4)
-#define decltype(expr) ::nbl::hlsl::type_traits::impl::decltype_t<typeid(expr)>::type
+#define typeid(expr) (sizeof(::nbl::hlsl::impl::encode_typeid(expr))/4)
+#define decltype(expr) ::nbl::hlsl::impl::decltype_t<typeid(expr)>::type
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:517]: oopsie forgot to start from tip of `type_traits` and not `master`
https://github.com/Devsh-Graphics-Programming/Nabla/pull/571

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 5]
```diff
@@ -59,8 +73,8 @@
 
   // type property queries
   template<class T> struct alignment_of; (TODO)
-  template<class T> struct rank; (TODO)
-  template<class T, unsigned I = 0> struct extent; (TODO)
+  template<class T> struct rank; (DONE)
+  template<class T, unsigned I = 0> struct extent; (DONE)
  
   // type relations
   template<class T, class U> struct is_same; (DONE)
```
[the remark this answers is not in the document]
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/type_traits.hlsl:80]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346540915

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 6]
```diff
@@ -34,7 +34,21 @@
...
+        int       is object
+        int&      is not an object
+        int*      is object
+        int*&     is not an object
+        cls       is object
+        cls&      is not an object
+        cls*      is object
+        int()     is not an object
+        int(*)()  is object
+        int(&)()  is not an object
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:50]: add comment "basically `!(is_reference || is_void || is_function)`"

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 7]
```diff
@@ -77,33 +91,33 @@
...
+  template<class T> struct remove_cv; (DONE)
+  template<class T> struct add_const; (DONE)
+  template<class T> struct add_volatile; (DONE)
+  template<class T> struct add_cv; (DONE)
 
   // reference modifications
-  template<class T> struct remove_reference; (TODO)
-  template<class T> struct add_lvalue_reference; (TODO)
-  template<class T> struct add_rvalue_reference; (TODO)
+  template<class T> struct remove_reference; (DONE)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:102]: by symmetry also `DONT USE` `remove_reference` (move it to `impl::`)

Comment (atkurtul) [reply include/nbl/builtin/hlsl/type_traits.hlsl:102]: i think this is safe to use though

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 8]
```diff
@@ -48,7 +62,7 @@
   template<class T> struct is_empty; (DONE? sizeof(T) == 0)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:62]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346099612 `standard_layout`

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 9]
```diff
@@ -77,33 +91,33 @@
...
-  template<class T> struct remove_reference; (TODO)
-  template<class T> struct add_lvalue_reference; (TODO)
-  template<class T> struct add_rvalue_reference; (TODO)
+  template<class T> struct remove_reference; (DONE)
+  template<class T> struct add_lvalue_reference; (DONE BUT DONT USE)
+  template<class T> struct add_rvalue_reference; (DONE BUT DONT USE)
 
   // sign modifications
   template<class T> struct make_signed; (TODO)
   template<class T> struct make_unsigned; (TODO)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:108]: update comment

File: include/nbl/builtin/hlsl/type_traits.hlsl [hunk 10]
```diff
@@ -309,40 +350,130 @@ template <class T>
...
+// dxc doesnt actually allow variables to be references
+// every reference is treated as having 'restrict' qualifier
+// https://godbolt.org/z/dsj99fY96
 namespace impl
 {
 
-template<uint32_t encoded_typeid>
-struct decltype_t;
+template<class>
+struct make_void { using type = void; };
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/type_traits.hlsl:397]: can be outside of `impl`, also list `void_t`  in the comments above
https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346115889
tell us to use `make_void` instead, and point out that unfortunately `make_void` is not variadic

**Pull request merged**